stream/replicator.h

```diff
@@ -403,15 +464,332 @@ struct StreamReplicatorImpl {
+    const auto flip_key = current::FromString<uint64_t>(r.url.query["key"]);
+
+    const auto head_idxts = stream_->Data()->HeadAndLastPublishedIndexAndTimestamp();
+    if (client_head > head_idxts.head) {
+      r("", HTTPResponseCode.BadRequest);
+      return;
+    }
+    if (has_i && (!Exists(head_idxts.idxts) || next_index > Value(head_idxts.idxts).index + 1)) {
+      r("", HTTPResponseCode.BadRequest);
+      return;
+    }
+    if (Exists(head_idxts.idxts)) {
+      const auto next_index_by_timestamp =
+          std::min(stream_->Data()->IndexRangeByTimestampRange(client_head + std::chrono::microseconds(1)).first,
+                   Value(head_idxts.idxts).index + 1);
+      if (next_index_by_timestamp != next_index) {
+        r("", HTTPResponseCode.BadRequest);
+        return;
+      }
+    }
+
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (!exposed_via_http_ || Exists(borrowed_publisher_)) {
+      r("", HTTPResponseCode.ServiceUnavailable);
+      return;
+    }
+    const auto& restrictions = exposed_via_http_->flip_restrictions_;
+    const auto max_clock_diff = restrictions.max_clock_diff_.count();
+    if (max_clock_diff > 0 && (!r.url.query.has("clock") ||
+                               std::abs(current::FromString<int64_t>(r.url.query["clock"]) -
+                                        current::time::Now().count()) > max_clock_diff)) {
+      r("", HTTPResponseCode.BadRequest);
+      return;
+    }
+    if (remote_follower_) {
+      try {
+        remote_follower_->PerformMasterFlip(*Value(stream_), flip_key);
+        remote_follower_ = nullptr;
+        stream_->BecomeMasterStream();
+      } catch (RemoteStreamRefusedFlipRequestException&) {
+        r("", HTTPResponseCode.BadRequest);
+        return;
+      }
+    } else if (flip_key != exposed_via_http_->flip_key_) {
+      r("", HTTPResponseCode.BadRequest);
+      return;
+    }
+
+    if (exposed_via_http_->flip_started_callback_) {
+      exposed_via_http_->flip_started_callback_();
+    }
+
+    const bool checked = r.url.query.has("checked");
+    // Grab the publisher to make sure no one can publish into it.
+    borrowed_publisher_ = stream_->BecomeFollowingStream();
+
+    // Send back the remaining diff.
+    const auto succeeded = ValidateAndSendDiff(r, next_index, client_head, checked);
+    if (!succeeded) {
+      // If something went wrong, restore the original state.
+      borrowed_publisher_ = nullptr;
+      stream_->BecomeMasterStream();
+      if (exposed_via_http_->flip_canceled_callback_) exposed_via_http_->flip_canceled_callback_();
+    } else if (exposed_via_http_->flip_finished_callback_) {
+      // Otherwise notify about flip completion from a separate thread
+      // to make it possible to unregister endpoints
+      // or even destroy the stream from inside the callback.
+      std::thread(exposed_via_http_->flip_finished_callback_).detach();
```

Comment:
Likely the cleanest solution for now, but could you add some comment re. `cron` or some `NextTick()` logic?

If scheduled callbacks with a thread pool would become a norm in Current, this place is a prime candidate to save on spawning a new thread, IMHO. If that would be happening, I'll `grep -R` in Current for at least `cron`, `-i nexttick` and `-i next_tick` :-)